# FEATURE  make req ssig message & partial block

blackironj opened the pull request:
#### 추가사항
[App] 
- AdminPlugin 에서 Login 시 app에 sk /cert /pass를 세팅 하도록 함.

[BlockProducer]
- MSG_REQ_SSIG에 sign하는 함수 추가
- MSG_REQ_SSIG를 만들 수 있도록 함수 추가
- MSG_REQ_SSIG와 MSG_BLOCK을 만드는데 사용하는 데이터가 중복이므로 Partial Block struct를 사용하도록 함.  ( MSG_REQ_SSIG 를 보낼 때 사용한 안에 block 데이터는 MSG_BLOCK을 만드는데 사용해야 함.)

#### 수정사항
[BlockProducer]
- TransactionPool을 chainplugin.hpp 에서 외부로 이동
- TransactionPool을 ChainPlugin을 통해 접근이 가능하도록 함

Commit:
# ADD - Application set/get pass, sk, cert

- AdminPlugin 에서 login 시 app에 sk / sk / cert를 세팅

Commit:
# FEATURE - [Block producer] make MSG_REQ_SSIG and partial block

 #### 추가사항
- MSG_REQ_SSIG 를 만들 수 있도록 함.
- MSG_REQ_SSIG와 MSG_BLOCK에서 사용하는 데이터가 중복 되므로
`PartialBlock` 을 만들도록함

 #### 수정사항
- TransactionPool을 chainplugin.hpp 에서 외부로 이동
- TransactionPool을 ChainPlugin을 통해 접근이 가능하도록 함

MoonSeongKwang commented on src/plugins/block_producer_plugin/block_producer_plugin.cpp:
```diff
@@ -219,6 +234,115 @@ class BlockProducerPluginImpl {
+
+    auto r_latest_block = chain.getBlocksByHeight(latest_resolved_block_height, latest_resolved_block_height);
+    auto csroot = r_latest_block[0].getContractStateRoot();
+    auto usroot = r_latest_block[0].getUserStateRoot();
```
이 csroot랑 usroot 부분에서 getBlocksByHeight로 여러 개의 블록을 불러와서 0번째를 선택하는 이유가 뭔가요? 그리고 getBlocksByHeight의 범위도 (latest_resolved_block_height, latest_resolved_block_height)로  똑같아 보이네요

blackironj replied:
csroot usroot는 현재만들고있는 block height - confirm level 의 height를 가진 block에서 가져오는거 같은데... 계산해보면 resolved 된 block중에 가장 최근이 아닌가요?
그래서 resolved pool에서 1개의 block만 가져오기 위해서 같은 height를 넣었구요.

MoonSeongKwang replied:
블록 하나만 가져오는 함수가 있는 줄 알았는데 높이만 가져오는 함수였네요. 그럼 그 함수 사용 자체는 이대로 하시면 될 것 같고,
일반적인 경우에는 latest resolved block이 되겠지만 명시적으로 confirm level만큼의 이전이라고 표기하는게 만약의 상황에서 더 좋을 것 같네요. 그런 상황이 어떤 식으로 벌어질 수 있을지 당장은 예시가 안 떠오르긴 하는데..

blackironj replied:
@SeongKwangMoon  
반드시 resolved 돼있을거라는 보장이 없단 말씀이시죠?

MoonSeongKwang replied:
방금 썼던 내용은 잘못 생각했네요. confirm level을 넘지 못하면 latest가 비어있을테니 결국 같은 행동이 되긴 하는데..

아무튼 만약 문제가 생긴다면 제일 긴쪽이 아닌 다른 더 낮은곳에 블록을 생성할 경우가 됩니다. 제 의견은 어차피 뺄셈 연산 한 번 더 하는게 큰 오버헤드가 아니니 명시적으로
`chain.getBlocksByHeight(latest_resolved_block_height, latest_resolved_block_height);`
를
`chain.getBlocksByHeight(height - BLOCK_CONFIRM_LEVEL, height - BLOCK_CONFIRM_LEVEL);` 로 하는 겁니다.

morishjs commented on src/plugins/admin_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -254,6 +254,10 @@ void LoginService::proceed() {
+        app().setPass(pass);
+        app().setSk(self_sk);
+        app().setMyCert(self_cert);
+
```
로그인 했을 때 3개의 함수가 호출되는데

rpc_services 에서는 app()의 인터페이스 3개를 알아야 한다는 의존성이 있어서, app() 내부에 인터페이스 하나로 추상화시켜서 내부적으로는 3개(private)를 호출하는게 좋아요. 
`completeUserSetup` 에 넣든가 아니면 다른 방법을 찾아봐야하는게 좋을 것 같네요

blackironj replied:
completeUserSignedIn 쪽에서 세팅하도록 수정할게요.

Commit:
# EDIT - Application set key & makePartialBlock

- Application 에서 key 관련 사항을 set함수를 제거하고
completeSignedIn 에서 값을 넣도록 수정

- BlockProducer의 makePartialBlock 에서 usroot txroot를 구할때 resolved
pool 에 block이 전혀 없을 경우도 처리하도록 수정

MoonSeongKwang commented on src/plugins/block_producer_plugin/block_producer_plugin.cpp:
```diff
@@ -251,11 +252,18 @@ class BlockProducerPluginImpl {
+      usroot = r_latest_block[0].getUserStateRoot();
+    } else {
+      array<uint8_t, 32> empty_bytes{0};
+      base64_type b64_empty_data = TypeConverter::encodeBase<64>(empty_bytes); //AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA1=
```
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA= 이긴 합니다.
